Changes to make integration with Paperwork possible

jflesch opened the pull request:
If you want, you can give it a try:

- ~~Changes in Paperwork are in the branch [wip-openpaperview](https://gitlab.gnome.org/World/OpenPaperwork/paperwork/-/tree/wip-openpaperview)~~ (Update: I've pushed my changes in Paperwork on the branch `master`. There is no branch `wip-openpaperview` anymore)
- Right now, only the CLI allows exposing the HTTPS server: `paperwork-cli https serve` will generate the certificates and start the HTTPS server. I'm working on the GUI.

I still have problems with Cache-Control. I'm unclear how it's supposed to work with your app:
Your app only specifies "Cache-Control: max-age=0" in its HTTP headers. I would have assumed it would send a header `If-Modified-Since` too, but it doesn't. So the server has no way to know if your app already got the latest version of the DB. I have a feeling you only get HTTP 304 because your app reuse the same HTTPS connection.

Also, I've spent the last 3 hours wondering why I got randomly some weird queries with URLs ending in "error". So I've allowed myself to remove your test code ...

Commit:
Add ON DELETE CASCADE on foreign keys.
Make the DB structure a bit more pratical for progressive updates in Paperwork.

Commit:
Sort documents by dates instead of rowid.
Rowid tell us the insertion order, not the document order.

Commit:
background.DownloadWorker: Remove random error injection.

This is really unexpected and disturbing for other devs.

jflesch commented:
The HTTPS server integrated in Paperwork is now announced on the local network too using mdns/zeroconf.

I've also added a pairing system using QRcodes to make the setup possibly easier. You can try it with `paperwork-cli https pair`.

The QRCodes provides all the data useful for a first safe HTTPS query. They all start with 3 bytes:
- 1 byte: pairing protocol version (always 0 for now)
- 1 byte: payload fragment index
- 1 byte: total number of payload fragment
- then: a payload fragment (binary)

When all the payload fragments are put side-by-side, you get a payload containing:
- 2 bytes indicating the length of the fingerprint of the HTTPS server certificate
- the SHA-256 fingerprint of the HTTPS server certificate
- 2 bytes indicating the length of a secret binary string
- a secret binary string
- 2 bytes indicating the TCP port number used by the HTTPS server
- all the known addresses of the HTTPS server (FQDN, zeroconf hostname, IP addresses, etc), separated by nul-characters (`\0`), by order of preference.

OpenPaperview could then try to query `GET https://<addr>:<port>/pair` with the header `Authorization: <the secret>`. It would then get a JSON payload provide the server certificate (X509), and a client certificate and secret key. For security, it would have to check the HTTPS server certificate against its fingerprint.

Feel free to tell me if you think anything should be changed/improved.

bwt commented:
Room is quite picky when it comes to schema modifications.
I will add the migration.


The cache for images and thumbnails is handled by Coil.

The DB cache is handled by OkHttp, but the max age is set to 0, in order to force a conditional GET.

In my experience the response's "max-age" value depends on the server configuration and is sometimes quite large.

OkHttp recommends setting the request's max age to 0 to make the server validate the cache content.
It uses an "If-None-Match" header or "If-Modified-Since" depending on the cached response



Also I would prefer a QR Code containing something more readable, like a JSON object.

jflesch commented:
I was actually returning HTTP/1.0 replies instead of 1.1. I've also added the ETag header. But even after I've fixed that, I'm still not getting `If-None-Match` nor `If-Modified-Since`. I've also tried adding `Cache-Control: max-age=604800`.

When I run the test integrated to your app (very handy by the way ! :), I get:

```
[INFO  ] [openpaperwork_core.http.server] GET /papers.sqlite
[INFO  ] [paperwork_backend.http        ] GET /papers.sqlite (Cache-Control=max-age=0, Host=deltang.home:12346, Connection=Keep-Alive, Accept-Encoding=gzip, 
User-Agent=okhttp/4.12.0)
[ERROR ] [openpaperwork_core.http.server] Missing client certificate !
192.168.42.203 - - [03/Oct/2024 23:51:16] "GET /papers.sqlite HTTP/1.1" 403 -


[INFO  ] [openpaperwork_core.http.server] GET /papers.sqlite
[INFO  ] [paperwork_backend.http        ] GET /papers.sqlite (Cache-Control=max-age=0, Host=deltang.home:12346, Connection=Keep-Alive, Accept-Encoding=gzip, 
User-Agent=okhttp/4.12.0)
[INFO  ] [paperwork_backend.http        ] HTTP 200: OK: BackendReply(http_code=200, content_type=None, content_length=37527552, last_modified=1727992270, 
etag=41241650040042186793163642098742452824809244748245334500777843548138945653721, content_fd=<openpaperwork_gtk.fs.gio._GioFileAdapter object at 
0x7f244ffd2e30>, content_raw=None)
192.168.42.203 - - [03/Oct/2024 23:51:17] "GET /papers.sqlite HTTP/1.1" 200 -


[INFO  ] [openpaperwork_core.http.server] GET /papers.sqlite
[INFO  ] [paperwork_backend.http        ] GET /papers.sqlite (Cache-Control=max-age=0, Host=deltang.home:12346, Connection=Keep-Alive, Accept-Encoding=gzip, 
User-Agent=okhttp/4.12.0)
[INFO  ] [paperwork_backend.http        ] HTTP 200: OK: BackendReply(http_code=200, content_type=None, content_length=37527552, last_modified=1727992270, 
etag=41241650040042186793163642098742452824809244748245334500777843548138945653721, content_fd=<openpaperwork_gtk.fs.gio._GioFileAdapter object at 
0x7f244ff1e0b0>, content_raw=None)
192.168.42.203 - - [03/Oct/2024 23:51:25] "GET /papers.sqlite HTTP/1.1" 200 -
```

bwt commented:
I don't understand either why the cache does not work. I need to setup paperwork from the sources and do tests.

A few remarks :

As I understand from [Force a Network Response](https://square.github.io/okhttp/3.x/okhttp/index.html?okhttp3/Cache.html), 
setting the cache control to "max-age=0" does not ignore the cache (as would "no-cache") but forces a conditional GET (instead of using the cache content without even making a network request), which is exactly what we want.

OkHttp seems to prefer "ETag" / "If-None-Match" over "Last-Modified" / "If-Modified-Since". Your response has both, so this shouldn't be the problem.

By the way when compiled in debug mode, the requests and responses are logged with all the headers.

bwt commented:
I tried with the current version of paperwork (commit 1288353)
and the results are very strange

It seems like the server sends the database once and then systematically responds "not modified" until it is restarted, regardless of the contents of the request.

Restarting the app has no effect.
Also removing the "max-age=0" from the requests doesn't change this behaviour (but it breaks the test because the second query is skipped by the client)


### The first run goes as expected : 200 then 304

#### Server's log (without the 403 query):

```
[INFO  ] [openpaperwork_core.http.server] GET /papers.sqlite
[INFO  ] [paperwork_backend.http        ] GET /papers.sqlite (Cache-Control=max-age=0, Host=192.168.31.80:12346, Connection=Keep-Alive, Accept-Encoding=gzip, User-Agent=okhttp/4.12.0)
[INFO  ] [paperwork_backend.http        ] HTTP 200: OK: BackendReply(http_code=200, content_type=None, content_length=6942720, last_modified=1728137870, etag=78150409426323468554337468744635238340913434436617443779310801157291418597740, 
content_fd=<openpaperwork_gtk.fs.gio._GioFileAdapter object at 0x7f9167c8ece0>, content_raw=None)
192.168.31.169 - - [05/Oct/2024 16:18:01] "GET /papers.sqlite HTTP/1.1" 200 -

[INFO  ] [openpaperwork_core.http.server] GET /papers.sqlite
[INFO  ] [paperwork_backend.http        ] GET /papers.sqlite (Cache-Control=max-age=0, Host=192.168.31.80:12346, Connection=Keep-Alive, Accept-Encoding=gzip, User-Agent=okhttp/4.12.0, 
If-None-Match="78150409426323468554337468744635238340913434436617443779310801157291418597740")
[INFO  ] [paperwork_backend.http        ] HTTP 304: Not Modified: BackendReply(http_code=304, content_type=None, content_length=6942720, last_modified=1728137870, etag=None, content_fd=None, content_raw=None)
192.168.31.169 - - [05/Oct/2024 16:18:01] "GET /papers.sqlite HTTP/1.1" 304 -
```

#### Client's log

```
--> GET https://192.168.31.80:12346/papers.sqlite http/1.1
Cache-Control: max-age=0
Host: 192.168.31.80:12346
Connection: Keep-Alive
Accept-Encoding: gzip
User-Agent: okhttp/4.12.0
--> END GET
<-- 200 OK https://192.168.31.80:12346/papers.sqlite (23ms)
Server: BaseHTTP/0.6 Python/3.12.6
Date: Sat, 05 Oct 2024 14:18:01 GMT
Content-Length: 6942720
Content-Type: application/octet-stream
Last-Modified: Sat, 05 Oct 2024 14:17:50 GMT
ETag: "78150409426323468554337468744635238340913434436617443779310801157291418597740"
Connection: close
<-- END HTTP
```

```
--> GET https://192.168.31.80:12346/papers.sqlite http/1.1
Cache-Control: max-age=0
Host: 192.168.31.80:12346
Connection: Keep-Alive
Accept-Encoding: gzip
User-Agent: okhttp/4.12.0
If-None-Match: 44635238340913434436617443779310801157291418597740"
--> END GET
<-- 304 Not Modified https://192.168.31.80:12346/papers.sqlite (5ms)
Server: BaseHTTP/0.6 Python/3.12.6
Date: Sat, 05 Oct 2024 14:18:01 GMT
Content-Length: 6942720
Content-Type: application/octet-stream
Last-Modified: Sat, 05 Oct 2024 14:17:50 GMT
Connection: close
<-- END HTTP
```

### For the following runs (until the server is restarted) the answer is always 304

#### Server's log :

```
[INFO  ] [openpaperwork_core.http.server] GET /papers.sqlite
[INFO  ] [paperwork_backend.http        ] GET /papers.sqlite (Cache-Control=max-age=0, Host=192.168.31.80:12346, Connection=Keep-Alive, Accept-Encoding=gzip, User-Agent=okhttp/4.12.0)
[INFO  ] [paperwork_backend.http        ] HTTP 304: Not Modified: BackendReply(http_code=304, content_type=None, content_length=6942720, last_modified=1728137870, etag=None, content_fd=None, content_raw=None)
192.168.31.169 - - [05/Oct/2024 16:20:15] "GET /papers.sqlite HTTP/1.1" 304 -

[INFO  ] [openpaperwork_core.http.server] GET /papers.sqlite
[INFO  ] [paperwork_backend.http        ] GET /papers.sqlite (Cache-Control=max-age=0, Host=192.168.31.80:12346, Connection=Keep-Alive, Accept-Encoding=gzip, User-Agent=okhttp/4.12.0)
[INFO  ] [paperwork_backend.http        ] HTTP 304: Not Modified: BackendReply(http_code=304, content_type=None, content_length=6942720, last_modified=1728137870, etag=None, content_fd=None, content_raw=None)
192.168.31.169 - - [05/Oct/2024 16:20:16] "GET /papers.sqlite HTTP/1.1" 304 -
```

#### Client's log

The server should not answer 304,
there is nothing in the request that allows it to know that the client's cache is up to date.

```
--> GET https://192.168.31.80:12346/papers.sqlite http/1.1
Cache-Control: max-age=0
Host: 192.168.31.80:12346
Connection: Keep-Alive
Accept-Encoding: gzip
User-Agent: okhttp/4.12.0
--> END GET
<-- 304 Not Modified https://192.168.31.80:12346/papers.sqlite (7ms)
Server: BaseHTTP/0.6 Python/3.12.6
Date: Sat, 05 Oct 2024 14:20:15 GMT
Content-Length: 6942720
Content-Type: application/octet-stream
Last-Modified: Sat, 05 Oct 2024 14:17:50 GMT
Connection: close
<-- END HTTP
```

```
--> GET https://192.168.31.80:12346/papers.sqlite http/1.1
Cache-Control: max-age=0
Host: 192.168.31.80:12346
Connection: Keep-Alive
Accept-Encoding: gzip
User-Agent: okhttp/4.12.0
--> END GET
<-- 304 Not Modified https://192.168.31.80:12346/papers.sqlite (6ms)
Server: BaseHTTP/0.6 Python/3.12.6
Date: Sat, 05 Oct 2024 14:20:16 GMT
Content-Length: 6942720
Content-Type: application/octet-stream
Last-Modified: Sat, 05 Oct 2024 14:17:50 GMT
Connection: close
<-- END HTTP
```

jflesch commented:
> It seems like the server sends the database once and then systematically responds "not modified" until it is restarted, regardless of the contents of the request.

My bad, sorry, I forgot to mention it's a workaround I've put in place for now. Otherwise OpenPaperview keeps updating the DB.

jflesch commented:
I've added the possibility to disable this workaround using an environment variable:
```
OPENPAPERVIEW_WORKAROUND_FORCE_CACHE_TIME=0 paperwork-cli https serve
```

bwt commented:
I'm still not able to reproduce the problem. In my case the client actually sends the "If-None-Match" header, but the server ignores it :

```
[INFO  ] [openpaperwork_core.http.server] GET /papers.sqlite
[INFO  ] [paperwork_backend.http        ] GET /papers.sqlite (Cache-Control=max-age=0, Host=192.168.31.80:12346, Connection=Keep-Alive, Accept-Encoding=gzip, User-Agent=okhttp/4.12.0)
[ERROR ] [openpaperwork_core.http.server] Missing client certificate !
192.168.31.169 - - [06/Oct/2024 22:57:36] "GET /papers.sqlite HTTP/1.1" 403 -

[INFO  ] [openpaperwork_core.http.server] GET /papers.sqlite
[INFO  ] [paperwork_backend.http        ] GET /papers.sqlite (Cache-Control=max-age=0, Host=192.168.31.80:12346, Connection=Keep-Alive, Accept-Encoding=gzip, User-Agent=okhttp/4.12.0)
[INFO  ] [paperwork_backend.http        ] HTTP 200: OK: BackendReply(http_code=200, content_type=None, content_length=6942720, last_modified=1728248250, etag=57525897707264668916749568462499791228998604059612589403086580994172393271869,
content_fd=<openpaperwork_gtk.fs.gio._GioFileAdapter object at 0x7f13cfca35b0>, content_raw=None)
192.168.31.169 - - [06/Oct/2024 22:57:36] "GET /papers.sqlite HTTP/1.1" 200 -

[INFO  ] [openpaperwork_core.http.server] GET /papers.sqlite
[INFO  ] [paperwork_backend.http        ] GET /papers.sqlite (Cache-Control=max-age=0, Host=192.168.31.80:12346, Connection=Keep-Alive, Accept-Encoding=gzip, User-Agent=okhttp/4.12.0,
If-None-Match="57525897707264668916749568462499791228998604059612589403086580994172393271869")
[INFO  ] [paperwork_backend.http        ] HTTP 200: OK: BackendReply(http_code=200, content_type=None, content_length=6942720, last_modified=1728248250, etag=57525897707264668916749568462499791228998604059612589403086580994172393271869,
content_fd=<openpaperwork_gtk.fs.gio._GioFileAdapter object at 0x7f13d2d81f60>, content_raw=None)
192.168.31.169 - - [06/Oct/2024 22:57:36] "GET /papers.sqlite HTTP/1.1" 200 -
```

So I could fix it by changing openpaperview.py :

```python
    def http_backend_get(self, request) -> Optional[paperwork_backend.http.BackendReply]:
        if request.path != QUERY_PATH:
            return None
        with DB_LOCK:

            etag = f'"{self.core.call_success("fs_hash", self.db_path)}"'
            if_none_match = request.headers["If-None-Match"]

            # content has not changed
            if if_none_match == etag:
                return paperwork_backend.http.BackendReply(
                    http_code=304,
                    last_modified=self.core.call_success("fs_get_mtime", self.db_path),
                    content_length=self.core.call_success("fs_getsize", self.db_path),
                    etag=etag,
                )

            ...
```

The result is now correct (403, 200, 304, even on successive runs)

```
[INFO  ] [openpaperwork_core.http.server] GET /papers.sqlite
[INFO  ] [paperwork_backend.http        ] GET /papers.sqlite (Cache-Control=max-age=0, Host=192.168.31.80:12346, Connection=Keep-Alive, Accept-Encoding=gzip, User-Agent=okhttp/4.12.0)
[ERROR ] [openpaperwork_core.http.server] Missing client certificate !
192.168.31.169 - - [06/Oct/2024 23:44:46] "GET /papers.sqlite HTTP/1.1" 403 -

[INFO  ] [openpaperwork_core.http.server] GET /papers.sqlite
[INFO  ] [paperwork_backend.http        ] GET /papers.sqlite (Cache-Control=max-age=0, Host=192.168.31.80:12346, Connection=Keep-Alive, Accept-Encoding=gzip, User-Agent=okhttp/4.12.0)
[INFO  ] [paperwork_backend.http        ] HTTP 200: OK: BackendReply(http_code=200, content_type=None, content_length=6942720, last_modified=1728251061, etag=89341580819238072844050355004032466107837520317825244966778922280465815396961,
content_fd=<openpaperwork_gtk.fs.gio._GioFileAdapter object at 0x7fbc1d8ab2e0>, content_raw=None)

192.168.31.169 - - [06/Oct/2024 23:44:47] "GET /papers.sqlite HTTP/1.1" 200 -
[INFO  ] [openpaperwork_core.http.server] GET /papers.sqlite
[INFO  ] [paperwork_backend.http        ] GET /papers.sqlite (Cache-Control=max-age=0, Host=192.168.31.80:12346, Connection=Keep-Alive, Accept-Encoding=gzip, User-Agent=okhttp/4.12.0,
If-None-Match="89341580819238072844050355004032466107837520317825244966778922280465815396961")
[INFO  ] [paperwork_backend.http        ] HTTP 304: Not Modified: BackendReply(http_code=304, content_type=None, content_length=6942720, last_modified=1728251061, etag='"89341580819238072844050355004032466107837520317825244966778922280465815396961"',
content_fd=None, content_raw=None)
192.168.31.169 - - [06/Oct/2024 23:44:47] "GET /papers.sqlite HTTP/1.1" 304 -
```

jflesch commented:
I did some tries. It seems somehow related to the sqlite database size.

My `papers.sqlite` makes about 38MB. With it, I don't get any `If-None-Match`.

But if I start Paperwork and then smash the DB with `echo toto > ~/.local/share/paperwork2/workdir_data/papers_cggL-9gZ/openpaperview.sqlite`, then it works.

Commit:
Increase OkHTTP cache size: papers.sqlite can be bigger than 32MB.

With 3700 documents, my papers.sqlite is about 40MB.
Some people have up to 11000 documents -> we can estimate their papers.sqlite will be about 128MB.

jflesch commented:
OK, found it :)

The OkHttp cache size was set to 32MB. Too small for my 40MB database. I've increased it to 256MB so we have some margin (I've 3700 documents, [some people appear to have up to 11800 documents](https://www.openpaper.work/fr/paperwork/statistics_pie) -> we can assume their DB will be about 128MB)

jflesch commented:
I've also removed the workaround in Paperwork and implemented Etag+If-None-Match properly (including on other files)

bwt commented on app/schemas/net.phbwt.paperwork.data.AppDatabase/1.json:
This file should not be updated, instead the version in the net.phbwt.paperwork.data.AppDatabase should be incremented, and the generated 2.json should be added. 
The version on the generated DB should also be changed accordingly  (`PRAGMA user_version = 2;`)

jflesch replied:
Done

bwt commented:
Good catch !

Can you revert the change in v1 of the DB et create a v2 ?

jflesch commented:
[Done](https://gitlab.gnome.org/World/OpenPaperwork/paperwork/-/commit/32abdd36958acd752287226e2efae33c322268a0)

bwt commented on tools/create_viewer_cb.py:
```diff
@@ -246,14 +247,15 @@ def create_db():
   main,
-  additional
+  additional,
+  CONSTRAINT fkDocument FOREIGN KEY (rowid) REFERENCES Document(documentId) ON DELETE CASCADE
```
Having a constraint here is not a good idea, specially with a cascade delete on a FTS table with external content
(in fact I'm surprised it's even possible)

jflesch replied:
Actually it's supported by Sqlite, but not by Room. I forgot to remove it here.

jflesch replied:
Fixed

bwt commented:
Sorry for being so slow to review the PR.

This is in fact more a suggestion for Paperwork on how handle the FTS tables.

The FTS is split in 2 parts : the index in `DocumentFts` and the content in `DocumentText`.

For them to remain in sync, modifications are a bit more complicated than ordinary tables.
For example a row must be deleted from `DocumentFts` first, then from `DocumentText`, so that SQLite can find the data that needs to be de-indexed.
Likewise, when the content is updated, the old value must be de-indexed, before the new value is re-indexed.

The recommended way is to use triggers to handle the FTS part.
In our case : 

```SQL
-- insert : index
CREATE TRIGGER DocumentText_ai AFTER INSERT ON DocumentText BEGIN
  INSERT INTO DocumentFts(docid, main, additional) VALUES(new.rowid, new.main, new.additional);
END;

-- update, de-index old value
CREATE TRIGGER DocumentText_bu BEFORE UPDATE ON DocumentText BEGIN
  DELETE FROM DocumentFts WHERE docid = old.rowid;
END;

-- update, re-index new value
CREATE TRIGGER DocumentText_au AFTER UPDATE ON DocumentText BEGIN
  INSERT INTO DocumentFts(docid, main, additional) VALUES(new.rowid, new.main, new.additional);
END;

-- delete, de-index
CREATE TRIGGER DocumentText_bd BEFORE DELETE ON DocumentText BEGIN
  DELETE FROM DocumentFts WHERE docid = old.rowid;
END;

```

The triggers could be useful in your case, but they should not be added to the `create_viewer_cb.py` script.
Depending on the configuration, the text is sometimes only indexed, without being stored.

Fortunately Room doesn't care about triggers when checking the schema.

For more details see [External Content FTS4 Tables](https://www.sqlite.org/fts3.html#_external_content_fts4_tables_)

jflesch commented:
Thanks for the suggestion, but don't worry, it's already handled on Paperwork side :). As said, I just forgot to remove this change in create_viewer_cb.py.

bwt commented:
Thank you. I will now work on the pairing.

Just ou of curiosity, you don't appear in the project's contributors.
Did you somehow choose to make your contribution anonymous ?

jflesch commented:
> Thank you.

You're welcome :)


> I will now work on the pairing.

Just beware, I haven't tested it much yet. I wouldn't be surprised if there are a bunch of bugs in it.


> Just ou of curiosity, you don't appear in the project's contributors.

Hu ? Where is it ?

Anyway, doesn't matter much I guess. It's very small contribution anyway.
